### ISSUE TITLE
Writing R code in a DCF is unappealing

### ISSUE THREAD

**Issue by hadley** [r-lib/lintr#1210]
Since you don't get syntax highlighting, help, autocomplete, ...

What do you think about supporting some additional way of defining custom linters that uses a file with a `.R` extension?

**Comment by AshesITR** [comment]
Rather minimal changes to the `.lintr` file would make it parseable R code, though. I'd be okay with allowing `.lintr` to also (in addition to being dcf-formatted) be parseable R code, restricted to assigning the lintr config variables, i.e., for our `.lintr`

```r
linters <- linters_with_defaults( # The following TODOs are part of an effort to have {lintr} lint-free (#584)
   line_length_linter = line_length_linter(120),
   # TODO add implicit_integer_linter()
   cyclocomp_linter = cyclocomp_linter(37), # TODO reduce to 15
)
exclusions <- list(
  "inst/doc/creating_linters.R" = 1,
  "inst/example/bad.R",
  "tests/testthat/default_linter_testcode.R",
  "tests/testthat/dummy_packages",
  "tests/testthat/dummy_projects",
  "tests/testthat/exclusions-test",
  "tests/testthat/knitr_formats",
  "tests/testthat/knitr_malformed"
)
```

@MichaelChirico WDYT?

**Comment by hadley** [comment]
Allowing `.lintr` to be either DCF or R doesn't seem that appealing to me either. I was thinking more of something like `_lintr.R`

There's some related art in Roxygen2 — you can configure simple options in `DESCRIPTION` or provide `man/roxygen/meta.R` if you want to run R code.

**Comment by AshesITR** [comment]
Not a big fan of allowing two separate configuration files - what to do if both files exist in the same directory?
`pkgdown`, `blogdown`, `covr` and GitHub Actions use YAML for configuration, by the way. I've never heard of the `man/roxygen/meta.R` feature.

I must admit though, that the DCF format took a bit to get used to, especially the akward indentation requirements.

**Comment by AshesITR** [comment]
Hmm that seems hard because we intend on `linters` being configured via `linters_with_defaults()` or `linters_with_tags()`.
For a transitional period, I'd be fine if `.lintr` would be parsed as YAML and, if that fails, as DCF with a deprecation warning.
I guess the question is, can we avoid R code by leveraging the YAML structure, e.g. using

```yaml
linters:
  - with: defaults
    line_length_linter:
      length: 120
    cyclocomp_linter:
      limit: 37
exclusions:
 - inst/doc/creating_linters.R: 1
 - inst/example/bad.R
 - tests/testthat/default_linter_testcode.R
 - tests/testthat/dummy_packages
 - tests/testthat/dummy_projects
 - tests/testthat/exclusions-test
 - tests/testthat/knitr_formats
 - tests/testthat/knitr_malformed
```

**Comment by MichaelChirico** [comment]
I also find the DCF approach a bit awkward... what's the appeal of it besides back-compatibility?

**Comment by AshesITR** [comment]
The only other advantage I can think of is `read.dcf()` is in base R.

**Comment by MichaelChirico** [comment]
right... that basically maps to the current setup though right? DCF keys/values --> variable names/values in config environment

**Comment by AshesITR** [comment]
True. But in `R` code it's more tempting to define other variables.

**Comment by hadley** [comment]
If you were worried about that you could inspect the environment and error if there were unexpected variables.

PS. I'd suggest `sys.source(".lintr", env = new.env(baseenv()))` instead of using `source()`
PPS. I'd also suggest using `_lintr.R` or similar instead of reusing the existing file name

**Comment by MichaelChirico** [comment]
if we're going to switch we could also offer a script (maybe a snippet in NEWS) to convert an existing DCF->R

**Comment by MichaelChirico** [comment]
I guess this would supersede #886

**Comment by AshesITR** [comment]
If we go with R, we should export a `lintr_config()` or `lintr_settings()` function that has all configuration options as arguments and does some validation.

**Comment by HaHeho** [comment]
I'm just using `lintr` for the first time. It's a great tool, but oh boy, is DCF terrible (since it is very easy to break).

Finding an alternative way for configuration is very desirable. The `.lintr.R` way might be fine. However, in the current state, this seems to suffer from similar problems due to the underlying DCF framework. E.g., impractical "malformatting" by not having the closing brackets indented (see [here](https://github.com/r-lib/lintr/issues/1530#issuecomment-1238351621)) still occurs with the R syntax. This formatting requirement should be removed (for good) since it does not exist in R itself.

**Comment by MichaelChirico** [comment]
@HaHeho Could you share an R file that's giving you this issue (in a new bug report please)?

.lintr.R is plain R code, completely independent of DCF implementation, so there should be no such issue.

**Comment by HaHeho** [comment]
> @HaHeho Could you share an R file that's giving you this issue (in a new bug report please)?
> 
> .lintr.R is plain R code, completely independent of DCF implementation, so there should be no such issue.

Hmm, maybe my R Studio session was borked then. I started with `.lintr` DCF and ran some linting. I then found the release notes on the experimental feature, renamed my file to `.lintr.R`, and adjusted the syntax. The linter recognized the file. However, it still complained about an invalid DCF format due to the indentation.

I did some more testing now by repeating the above process. I learned that executing `lintr::lint("file")` in the console works fine, but executing the adding via a keyboard shortcut causes the following error.
```
> lintr:::addin_lint()
Error in read.dcf(config_file, all = TRUE) : Invalid DCF format.
Regular lines must have a tag.
Offending lines start with:
  linters <- linters_with_defaults(
  )
```
The `.lintr.R` file is an almost empty:
```
linters <- linters_with_defaults(
  object_name_linter = NULL
)
```

Is this a known issue that can be fixed easily? Should I file a bug report for this one?

**Comment by MichaelChirico** [comment]
> Should I file a bug report for this one?

Indeed it looks like a bug, thanks! This `read.dcf()` usage needs to be updated:

https://github.com/r-lib/lintr/blob/106dfb4036a63d8bbe211b0989ad09723fd0f312/R/addins.R#L15

Please file a new bug.
